runtime/etc/Makefile.mli-shared

```diff
@@ -0,0 +1,140 @@
+
+all: checkZMQ buildServer buildClient runTagsCommand FORCE
+
+checkZMQ:
+ifndef CHPL_ZMQ_LOC
+  $(error CHPL_ZMQ_LOC must be set to the location of the ZeroMQ library)
+endif
+
+buildServer: $(TMPSERVERNAME)
+
+buildClient: $(TMPBINNAME)
+
+runTagsCommand:
+	$(TAGS_COMMAND)
+
+$(TMPBINNAME): $(CHPL_CL_OBJS) FORCE
+#
+# Invoke the launcher script, but don't actually link up the launcher binary.
+# Do this by setting CHPL_MAKE_IS_MLI=1.
+#
+	$(MAKE) -f $(CHPL_MAKE_HOME)/runtime/etc/Makefile.launcher all \
+		CHPL_MAKE_HOME=$(CHPL_MAKE_HOME) \
+		TMPBINNAME=$(TMPSERVERNAME) \
+		BINNAME=$(SERVERNAME) \
+		TMPDIRNAME=$(TMPDIRNAME) \
+		CHPL_MAKE_RUNTIME_LIB=$(CHPL_MAKE_RUNTIME_LIB) \
+		CHPL_MAKE_RUNTIME_INCL=$(CHPL_MAKE_RUNTIME_INCL) \
+		CHPL_MAKE_THIRD_PARTY=$(CHPL_MAKE_THIRD_PARTY) \
+		CHPL_MAKE_IS_MLI=1
+#
+# Compile the client bundle. Exactly the same as for static compilation, the
+# only thing that changes here are the flags inserted by codegen_makefile.
+#
+	$(CC) $(CHPL_MAKE_BASE_CFLAGS) $(GEN_CFLAGS) $(COMP_GEN_CLIENT_CFLAGS) \
+		-c -o $(TMPBINNAME).o \
+		$(CHPLSRC) \
+		$(COMP_GEN_MLI_EXTRA_INCLUDES) \
+		$(CHPL_RT_INC_DIR) \
+		$(CHPL_MLI_INCS)
+#
+# Invoke the linker directly rather than package everything into an archive.
+# Again, the codegen_makefile function is responsible for inserting the
+# "shared library" flag here.
+#
+	$(LD) $(CHPL_MAKE_BASE_LFLAGS) $(GEN_LFLAGS) $(COMP_GEN_CLIENT_LFLAGS) \
+		-o $(TMPBINNAME) \
+		$(TMPBINNAME).o $(TMPSERVERNAME)_launcher.o \
+		$(CHPLUSEROBJ) $(CHPL_CL_OBJS) \
+		-L$(CHPL_LN_LIB_DIR) \
+		-lm -lchpllaunch \
+		$(LAUNCH_LIBS) \
+		-L$(CHPL_ZMQ_LOC)/lib \
+		-lzmq
+#
+# Last minute munging in case of output file name collisions.
+#
+ifneq ($(TMPBINNAME),$(BINNAME))
+	cp $(TMPBINNAME) $(BINNAME)
+	rm $(TMPBINNAME)
+endif
+#
+# On Darwin, we need to adjust the "install_name" of the final shared library
+# file so that it points to $(BINNAME) and not $(TMPBINNAME).
+#
+# TODO: Consider moving this into the compiler? Might be less duplication
+# across all the Makefiles.
+#
+ifeq ($(UNAME),Darwin)
+	install_name_tool $(BINNAME) -id "$(BINNAME)"
+endif
+
+$(TMPSERVERNAME): $(CHPL_CL_OBJS) checkRtLibDir FORCE
+ifneq ($(SKIP_COMPILE_LINK),skip)
+#
+# Compile the server bundle.
+#
+	$(CC) $(CHPL_MAKE_BASE_CFLAGS) $(GEN_CFLAGS) $(COMP_GEN_SERVER_CFLAGS) \
+		-c -o $(TMPSERVERNAME).o \
+		$(CHPLSERVERSRC) \
+		$(CHPL_RT_INC_DIR) \
+		$(CHPL_MLI_INCS) \
+		$(COMP_GEN_MLI_EXTRA_INCLUDES)
+#
+# Compile loose source files in CHPLUSEROBJ.
+#
+	$(foreach srcFile, $(CHPLUSEROBJ), \
+		$(CC) $(CHPL_MAKE_BASE_CFLAGS) $(GEN_CFLAGS) $(COMP_GEN_SERVER_CFLAGS) \
+			-c -o $(srcFile) $(CHPL_RT_INC_DIR) $(srcFile).c \
+	;)
+#
+# Link the Chapel runtime and the ZMQ library together into a multi-locale
+# executable.
+#
+# Just a quick note to self:
+#
+#   - $(LIBS) contains gasnet stuff (gasnet link errors otherwise).
+#   - $(CHPL_MAKE_THIRD_PARTY_LINK_ARGS) is everything, including libchpl.
```

Comment:
More precisely, `$(CHPL_MAKE_THIRD_PARTY_LINK_ARGS)` is all the third-party libraries other than the gasnet-related ones.  It does include `-lchpl`, but note that we also specify `-lchpl` explicitly in the link line below, right after `-L$(CHPL_RT_LIB_DIR)`, so it's probably not necessary to mention that it's hidden in `$(CHPL_MAKE_THIRD_PARTY_LINK_ARGS)` as well.